EIPS/eip-3540.md

```diff
@@ -93,88 +90,96 @@ The list of section headers is terminated with the *section headers terminator b
+
+(*) For not yet deployed containers this can be greater than the actual content length.
+
+#### Body
+
+| name              | length   | value         | description                                                                                |
+|-------------------|----------|---------------|--------------------------------------------------------------------------------------------|
+| types_section     | variable | n/a           | stores code section metadata                                                               |
+| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                         |
+| outputs           | 1 byte   | 0x00-0x7F     | number of stack elements the code section returns                                          |
+| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the operand stack by the code section          |
+| code_section      | variable | n/a           | arbitrary bytecode                                                                         |
+| container_section | variable | n/a           | arbitrary EOF-formatted container                                                          |
+| data_section      | variable | n/a           | arbitrary sequence of bytes                                                                |
 
 See [EIP-4750](./eip-4750.md) for more information on the type section content.
 
+**NOTE**: A special value of `outputs` being `0x80` is designated to denote non-returning functions as defined in a separate EIP.
+
 #### EOF version 1 validation rules
 
-1. In addition to general validation rules above, EOF version 1 bytecode conforms to the rules specified below:
-   - Exactly one type section header MUST be present immediately following the EOF version. Each code section MUST have a specified type signature in the type body.
-   - Exactly one code section header MUST be present immediately following the type section. A maximum of 1024 individual code sections are allowed.
-   - Exactly one data section header MUST be present immediately following the code section.
-2. Any version other than `0x01` is invalid.
+The following validity constraints are placed on the container format:
 
-(*Remark:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+- `version` must be `0x01`
+- `types_size` is divisible by `4`
+- the number of code sections must be equal to `types_size / 4`
+- the number of code sections must be greater than `0` and not exceed `1024`
+- `code_size` may not be `0`
+- the number of container sections must not exceed `256`. The number of container sections may not be `0`, if declared in the header
+- `container_size` may not be `0`
+- data section is mandatory, but `data_size` may be `0`
+- data body length may be shorter than `data_size` for a not yet deployed container
 
```

Comment:
This list is different than [Container Validation](https://github.com/ipsilon/eof/blob/main/spec/eof.md#container-validation)